### PR TITLE
Updated surveys Branch

### PR DESCRIPTION
**Pull request by sohilpandya**
This is an offshoot of the #286.

------------------------------------
Content by **Franz**
PR related to #263 

#### Lecturer
- **queries**:
  - [x] new schema
  - [x] create survey (nullify all correct answers that may be sent from front-end)
  - [x] get surveys
  - [ ] get survey
  - [ ] edit survey (back-end adaptation)
  - [x] save student response to survey question
- **components / containers / actions**:
  - surveys recap: 
    - [x] lecturer (adapted from quizzes recap)
  - [x] get surveys on-enter 
  - [x] create survey (modify create quiz)
  - [ ] view survey (needs route, component adaptation, pass down to server, isSurvey and id)
  - [ ] edit survey  (needs route, component adaptation, pass down to server, isSurvey and id)

### CODE REVIEW

**Comment by sohilpandya** [comment]
@Danwhy @shouston3 I've had to rebase and then fix the conflicts again. But this branch is ready to go, contains all the changes Franz has done 👍

File: test/back/db/saveStudentResponse.test.js [hunk 1]
```diff
@@ -9,17 +9,21 @@ test('`saveStudentResponse` works', (t) => {
...
+        const survey_id = null;
         const question_id = 1;
         const response = 'c';
 
-        saveStudentResponse(pool, user_id, quiz_id, question_id, response, (error, response) => {
-            t.equal(error, expectedError, 'error is null, response is saved to db correctly.');
-            t.deepEqual(response.command, expectedCommand, 'Correct command of INSERT, response is saved to db correctly');
+        saveStudentResponse(pool, user_id, quiz_id, survey_id, question_id, response, (error, response) => {
+            if (error) {
+                t.error('should not have errored');
```
Comment (shouston3) [inline test/back/db/saveStudentResponse.test.js]: If we error, I think our tests may may hang. If this is the case, it probably isn't desirable

File: test/back/endpoints/endpoints.test.js [hunk 2]
```diff
@@ -141,7 +141,7 @@ test('/ endpoint works returns the correct payload', (t) => {
...
+        expected: { medals: { condition: [39, 69], medal_name: ['bronze', 'silver', 'gold'] }, module_id: 'TEST', name: 'test module', num_enrolled: 5, quizzes: [{ is_last_quiz: false, is_presented: true, name: 'Week 1 Quiz', num_entries: '4', num_questions: '2', quiz_id: 1 }, { is_last_quiz: false, is_presented: true, name: 'Week 2 Quiz', num_entries: '3', num_questions: '3', quiz_id: 2 }], surveys: [{ is_presented: true, name: 'Week 1 Survey', num_entries: '4', num_questions: '2', survey_id: 1 }, { is_presented: true, name: 'Week 2 Survey', num_entries: '4', num_questions: '2', survey_id: 2 }], trophies: { condition: [1, 100, 65, 2], trophy_name: ['first_quiz', 'high_score', 'overall_average', 'participation'] } }
```
Comment (shouston3) [inline test/back/endpoints/endpoints.test.js]: One line! I can't read this

**Comment by sohilpandya** [comment]
@shouston3 no, they are not done, this is just all the work that Franz had done. 

I've also addressed your comments above.

**Comment by shouston3** [comment]
I found this pr **incredibly** hard to read.

This is absolutely in no way the fault of @sohilpandya or @franzmoro.

I feel the reason is that we are not addressing the code quality issues and I think the longer they take to be addressed, the longer it will take to get readable pr's and higher the likelihood of mistakes to be made.

Currently, every time we need to manage a new piece of data we are creating a whole new set of database functions and queries and an entirely new endpoint (which is terrible), causing the codebase to bloat and in doing so, gives absolutely no context for the reviewer to see how the code is changing. (If we instead modify existing (more flexible functions), then diffs will be much easier).

It results in repetitive code which will take longer to write, will be much harder to review and be far more difficult to maintain. This is not the fault of @sohilpandya or @franzmoro, but a result of the current architecture.

Also, this pr is doing way too much (again this is in no way the fault of either @sohilpandya or @franzmoro, it's down to code quality - I feel that @franzmoro tried to alter syntax, architecture and functionality all at the same time), see: https://github.com/dwyl/contributing#pull-requests for why this is bad, but like me @franzmoro likes to address problems as they arise, I've had to hold back for #323, but would really like the accelerator to be put on with resolving this fundamental issue.

**Note for future**: Whenever there are bullet points in a pr, make them separate pr's.

**Readability note**: I consider #323 to be quite a large pr in terms of lines changed, but despite it being only around half the number of lines changed as this pr, it is much easier to read and only addresses one issue.

Thoughts @Danwhy

**Comment by shouston3** [comment]
I found this pr incredibly hard to read.

This is absolutely in no way the fault of @sohilpandya or @franzmoro.

I feel the reason is that we are not addressing architectural issues and I think the longer they take to be addressed, the longer it will take to get readable pr's and higher the likelihood of mistakes to be made.

Currently, every time we need to manage a new piece of data we are creating a whole new set of database functions and queries and an entirely new endpoint (which is terrible), causing the codebase to bloat and in doing so, gives absolutely no context for the reviewer to see how the code is changing. (If we instead modify existing (more flexible functions), then diffs will be much easier).

It results in repetitive code which will take longer to write, will be much harder to review and be far more difficult to maintain. This is not the fault of @sohilpandya or @franzmoro, but a result of the current architecture.

Also, this pr is doing way too much (again not the fault of either @sohilpandya or @franzmoro, it's down to code quality - I feel that @franzmoro tried to alter syntax, architecture and functionality all at the same time), see: https://github.com/dwyl/contributing#pull-requests for why this is bad, but like me @franzmoro likes to address problems as I see them, I've had to hold back for #323, but would really like the accelerator to be put on with resolving this fundamental issue.

**Note for future**: Whenever there are bullet points in a pr, make them separate pr's.

**Readability note**: I consider #323 to be quite a large pr in terms of lines changed, but despite it being only around half the number of lines changed as this pr, it is much easier to read and only addresses one issue.

Thoughts @Danwhy

**Comment by sohilpandya** [comment]
Thanks for the detailed review @shouston3, really appreciate it. I've got a couple of questions 

- Could you please pinpoint me as to what exactly you've found difficult to read? We can make a note of these areas and then create an issue so that it's in the backlog and can be looked at. 
- You refer to making multiple functions/queries to fetch information from a database, our original thoughts were that it was modular and easy to tests each individual function. Although would love for you to  your thoughts on how better to tackle this as you think the original approach is terrible. 

@Danwhy would love your thoughts as well

**Comment by shouston3** [comment]
@sohilpandya If I was assigned this issue, I would have ended up making a very similar pr, it was a really difficult task, the branches had diverged so much really great job getting it all to the same state.

Also, I’m not going to ask you to make any more amendments to this pr, if any are to be made in terms of readability (from @danwhy ’s comments) I’ll work on them.

(I have been guilty of some of this in my last pr which was also unreadable and this isn't anyones fault).

### Readability

* It's quite a case by case thing, but for example looking at:
 - `server/lib/getModuleForLecturer.js`, the callback chain is giving quite an unreadable diff, if we fix when a function does too much, then this will improve - again no ones fault, but the fault of not spending time addressing [callback hell](http://stackabuse.com/avoiding-callback-hell-in-node-js/)
 - `src/js/components/new-quiz/questions.js` so much refactoring went on in this file which should have been in a separate pull request. Mixing refactoring and functionality changes makes for a difficult to read pull request.
* As I said above, there were such a combination of issues addressed all at once in so many files that made it hard to read.

### Api issues

I am _so_ sorry for using the word terrible, this architecture was something that I have seen employed many times before (by myself in projects) and it was totally wrong to say.

It is really great to prioritise modularisation and testing 👍 , your hard work in doing so will make the process of refactoring much easier, I just feel that this current approach could scale better and I think that the options which allow for better scaling won't have to sacrifice modularisation and testing.

To clarify, I have seen projects where the current way of doing things has been great and readable and I would have no complaints, 4 or less callbacks are great and less than 20 files in a directory is fantastic, but we've now started to reach the 6th callback and reached 60+ files in `server/lib`. I feel that there should be time spent looking at the current functions and seeing if things can be simplified.

I feel that this project has recently started to see the pain of having isolated helpers for each database call (basically the `server/lib` is starting to grow a bit too big and all our handlers are growing in complexity when a lot of them are doing fairly similar things and should be linked), not doing anything about it and generating more similar functions will get unmanageable.

**Pull request merged**